compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1961,34 +2031,92 @@ object desugar {
+          case (Tuple(ts1), Tuple(ts2)) => ts1.corresponds(ts2)(deepEquals)
+          case _ => false
+
+      if betterForsEnabled then
+        enums match {
+          case Nil => body
+          case (gen: GenFrom) :: Nil =>
+            if gen.checkMode != GenCheckMode.Filtered // results of withFilter have the wrong type
+              && deepEquals(gen.pat, body)
+            then gen.expr  // avoid a redundant map with identity
+            else Apply(rhsSelect(gen, mapName), makeLambda(gen, body))
+          case (gen: GenFrom) :: rest
+          if rest.dropWhile(_.isInstanceOf[GenAlias]).headOption.forall(e => e.isInstanceOf[GenFrom]) =>
+            val cont = makeFor(mapName, flatMapName, rest, body)
+            val selectName =
+              if rest.exists(_.isInstanceOf[GenFrom]) then flatMapName
+              else mapName
+            Apply(rhsSelect(gen, selectName), makeLambda(gen, cont))
+          case (gen: GenFrom) :: (rest @ GenAlias(_, _) :: _) =>
+            val (valeqs, rest1) = rest.span(_.isInstanceOf[GenAlias])
+            val pats = valeqs map { case GenAlias(pat, _) => pat }
+            val rhss = valeqs map { case GenAlias(_, rhs) => rhs }
+            val (defpat0, id0) = makeIdPat(gen.pat)
+            val (defpats, ids) = (pats map makeIdPat).unzip
+            val pdefs = valeqs.lazyZip(defpats).lazyZip(rhss).map { (valeq, defpat, rhs) =>
+              val mods = defpat match
+                case defTree: DefTree => defTree.mods
+                case _ => Modifiers()
+              makePatDef(valeq, mods, defpat, rhs)
+            }
+            val rhs1 = makeFor(nme.map, nme.flatMap, GenFrom(defpat0, gen.expr, gen.checkMode) :: Nil, Block(pdefs, makeTuple(id0 :: ids)))
+            val allpats = gen.pat :: pats
+            val vfrom1 = GenFrom(makeTuple(allpats), rhs1, GenCheckMode.Ignore)
+            makeFor(mapName, flatMapName, vfrom1 :: rest1, body)
+          case (gen: GenFrom) :: test :: rest =>
+            val filtered = Apply(rhsSelect(gen, nme.withFilter), makeLambda(gen, test))
+            val genFrom = GenFrom(gen.pat, filtered, GenCheckMode.Filtered)
+            makeFor(mapName, flatMapName, genFrom :: rest, body)
+          case GenAlias(_, _) :: _ =>
+            val (valeqs, rest) = enums.span(_.isInstanceOf[GenAlias])
+            val pats = valeqs.map { case GenAlias(pat, _) => pat }
+            val rhss = valeqs.map { case GenAlias(_, rhs) => rhs }
+            val (defpats, ids) = pats.map(makeIdPat).unzip
+            val pdefs = valeqs.lazyZip(defpats).lazyZip(rhss).map { (valeq, defpat, rhs) =>
+              val mods = defpat match
+                case defTree: DefTree => defTree.mods
+                case _ => Modifiers()
+              makePatDef(valeq, mods, defpat, rhs)
+            }
+            Block(pdefs, makeFor(mapName, flatMapName, rest, body))
+          case _ =>
+            EmptyTree //may happen for erroneous input
+        }
+      else {
```

Comment:
There's a lot of duplication between the two branches, which is bad for two reasons:

 - it makes code maintenance harder since we have to fix potential issues twice
 - it makes it much harder to see what was changed by betterFors

Consider merging the two alternatives and applying more targetted deltas in the smallest parts where they are needed.